src/main/groovy/com/google/protobuf/gradle/ProtobufModelPlugin.groovy

```diff
@@ -0,0 +1,255 @@
+    }
+
+    @Mutate
+    void createDefaultProtobufSourceSets(ModelMap<NativeComponentSpec> components) {
+      components.all { component ->
+        if (!component.sources.containsKey('proto')) {
+          component.sources.create("proto", ProtobufSourceSet)
+        }
+      }
+    }
+
+    @Finalize
+    void bindDefaultProtoSourceSetToDefaultCppSourceSet(ModelMap<NativeComponentSpec> components) {
+      components.all {
+        sources.proto.generatedSourceSet.withType(CppSourceSet) {
+          sources.cpp.lib it
+        }
+      }
+    }
+
+    @Mutate
+    void configureGeneratedSourceSets(TaskContainer tasks, ModelMap<NativeBinarySpec> binaries,
+                                      @Path("buildDir") File buildDir,
+                                      ServiceRegistry serviceRegistry,
+                                      ProtobufConfigurator protobuf) {
+      final FileResolver fileResolver = serviceRegistry.get(FileResolver.class);
+
+      binaries.each { binary ->
+        binary.tasks.withType(GenerateProtoTask) { task ->
+          tasks.create("compile${task.name.capitalize()}Cpp", CppCompile) {
+            SourceDirectorySet source = new DefaultSourceDirectorySet("${task.name}Cpp", "${task.name.capitalize()}Cpp", fileResolver);
+            source.srcDir(new File("${task.outputBaseDir}/cpp"));
+            it.source(source);
+            it.includes(new File("${task.outputBaseDir}/cpp"));
+
+            it.setDescription(String.format("Compiles the %s of %s", source, binary));
+            it.setObjectFileDir(new File("${buildDir}/objs/${binary.namingScheme.outputDirectoryBase}/${task.name}Cpp"));
+
+            def cppCompiler = binary.cppCompiler;
+            it.setMacros(cppCompiler.getMacros());
+            it.setCompilerArgs(cppCompiler.getArgs());
+            it.setTargetPlatform(binary.getTargetPlatform());
+            it.setToolChain(binary.getToolChain());
+            it.includes(new Callable<List<FileCollection>>() {
+              public List<FileCollection> call() {
+                Collection<NativeDependencySet> libs = binary.getLibs();
+                return CollectionUtils.collect(libs, new Transformer<FileCollection, NativeDependencySet>() {
+                  public FileCollection transform(NativeDependencySet original) {
+                    return original.getIncludeRoots();
+                  }
+                });
+              }
+            });
+
+
+            it.dependsOn(task);
+
+            binary.tasks.add(it);
+            binary.builtBy(it);
+            binary.binaryInputs(it.getOutputs().getFiles().getAsFileTree().matching(new PatternSet().include("**/*.obj", "**/*.o")));
+          }
+        }
+      }
+    }
+  }
+
+  // TODO: This should probably be pushed into Gradle core.
+  public class CppFile implements TransformationFileType {}
+
+  private static class Proto implements LanguageTransform<ProtobufSourceSet, CppFile> {
```

Comment:
This is the only class that needs to use internal interface when using the Gradle mechanize for automatically creating the appropriate language transform task for the declared source set. Basically, it describe that all source from ProtobufSourceSet will be transformed into CppFile according to the specification of this class.

I'm keeping an eye out for finding a better way (not using internal class/interface) to achieve this transformation. It was discuss in these threads:
 * https://groups.google.com/d/topic/gradle-dev/l4KTozshWXA/discussion
 * https://groups.google.com/d/topic/gradle-dev/jteaSNnnFtY/discussion